HeterogeneousCore/CUDAUtilities/interface/prefixScan.h

```diff
@@ -0,0 +1,52 @@
+#ifndef HeterogeneousCore_CUDAUtilities_prefixScan_h
+#define HeterogeneousCore_CUDAUtilities_prefixScan_h
+
+#include <cstdint>
+#include <cassert>
+
+template<typename T>
+__device__
+void 
+__forceinline__
+warpPrefixScan(T * c, uint32_t i) {
+   auto x = c[i];
+   auto laneId = threadIdx.x & 0x1f;
+   #pragma unroll
+   for( int offset = 1 ; offset < 32 ; offset <<= 1 ) {
+     auto y = __shfl_up_sync(0xffffffff,x, offset);
+     if(laneId >= offset) x += y;
+   }
+   c[i] = x;
+}
+
+// limited to 32*32 elements....
+template<typename T>
+__device__
+void
+__forceinline__
+blockPrefixScan(T * c, uint32_t size, T* ws) {
```

Comment:
CUB seems to provide something similarly named
* `cub::WarpScan<...>` https://nvlabs.github.io/cub/classcub_1_1_warp_scan.html
* `cub::BlockScan<...>` https://nvlabs.github.io/cub/classcub_1_1_block_scan.html

Is there a reason for not using them (e.g. they do different thing, interface is awkward etc)?

Reply:
interface is a bit awkward, maybe I need to look at it again (now that I managed to use the host API)